Throw an error for parameter re-definitions across different blocks

danielzuegner opened the pull request:
### What does this implement/fix?
Throws an error when the same parameter is re-defined in any two of [`grid`, `fixed`, `random`]. Note that this does **not apply** when the parameter is re-defined in the **same block** (`grid`, `fixed`, or `random`) across levels.

 For example, the following **will** throw an error:
```yaml
fixed:
  param1: 123

sub_config:
  grid:
    param1:
      type: choice
      options:
        - 111
        - 222
```

while the following configuration will not:
```yaml
fixed:
  param1: 123

sub_config:
  fixed:
    param1: 222
```

Moreover, the error message indicates which parameters at which level are conflicting, e.g.:
```
ERROR: Parameters {'param1'} are defined in both 'grid' and 'fixed' in subconfig 'sub_config'.
ERROR: Found duplicate parameters in subconfig named 'sub_config'. Aborting.
```

klicperajo commented:
What is the reasoning behind throwing errors depending on the sampling method? I'd suggest always showing a warning and then choosing the one deeper in the hierarchy.

danielzuegner commented:
We can do that; my proposal additionally captures the case that the same parameter is re-defined on the same level, e.g. 
```yaml
grid:
  param1:
    type: choice
    options:
      - 1
      - 2

fixed:
  param1: 3
```

Do you think it's better to fix this separately?

klicperajo commented:
Okay, so the way I see it there are 2 cases.

a) Multiple definitions on the same level (either in different blocks or in the same block). In this case I agree that we should throw an error.
b) Multiple definitions on different levels. In this case I would show a warning and continue. We should make sure to only show this warning once, ideally listing all variables that we encountered more than once.

danielzuegner commented:
Hi, I've added a pretty large commit I've been working on for some time. 

As indicated by the commit message, these are the most important changes:

* Throw warnings for parameter redefinitions.
* Throw errors for duplicate parameters on the same level.
* Deprecate parameter collections in favor of dot-notation in dictionaries.
* Write tests for queuing and dictionary utility functions.
* Update the example and readme.
* Add advanced example showing how to use nested dictionaries and Sacred's prefixes.

The changes in the dictionary handling (dot-notation etc.) were a bit tricky to get right, which is why I've added numerous test cases to make sure the behavior is as desired. 

For repeated / duplicate parameters, I think the behavior is now as you described in your previous post: 
* parameter redefinition **on the same level**: throws error.
* parameter redefinition **across the hierarchy**: raises warning.

Let me know what you think!

klicperajo commented on examples/example_config.yaml:
```diff
@@ -68,28 +68,18 @@ slurm:
-  reg_scale: 0.0
-  keep_prob: 0.5
   max_epochs: 500
   patience: 10
```
Should we remove some hyperparameters that don't show anything new? Reducing this to the essentials should make the example easier to understand.

danielzuegner replied:
Fixed in a40eca8.

klicperajo commented on examples/example_experiment.py:
```diff
@@ -22,12 +22,12 @@ def config():
-def run(dataset, hidden_sizes, learning_rate, reg_scale, keep_prob, max_epochs, patience, display_step,
-        regularization_params):
+def run(dataset, hidden_sizes, learning_rate, max_epochs, patience, display_step,
+        regularization_params: dict):
```
If you add type annotations for one parameter, please add them for all.

klicperajo commented on seml/utils.py:
```diff
@@ -2,35 +2,102 @@
 
 
-def unflatten(dictionary: dict, sep: str = '.'):
+def unflatten(dictionary: dict, sep: str = '.', recursive: bool = False, levels=None):
```
`recursive=False` is the same as `levels=[0]`. Do we need both?

danielzuegner replied:
It's actually not the same. For example, unflattening 
```python
{
"a.b.c": 111,
"a.d": {"e.f": 222},
}
```
with `recursive=False, levels=None` will yield
```python
{
"a": {
  "b": {
    "c": 111,
  }
  "d": {"e.f": 222}
}
}
```
while `recursive=True, levels=None` will produce 
```python
{
"a": {
  "b": {
    "c": 111,
  }
  "d": {
    "e: {
      "f": 222
    }
  }
}
}
```

That is, `recursive=True` tells us that we also look into (potentially flattened) sub-dictionaries and also unflatten these.

klicperajo replied:
Ah, I misunderstood this then.

klicperajo reviewed (commented):
Added some comments for further improvements I noticed. :)

danielzuegner commented on examples/advanced_example_config.yaml:
```diff
@@ -0,0 +1,126 @@
+seml:
+  executable: examples/advanced_example_experiment.py
+  name: advanced_example_experiment
+  output_dir: examples/logs
```
I agree. See a40eca8.

danielzuegner commented on examples/README.md:
```diff
@@ -62,10 +55,10 @@ random:
-    type: randint
-    min: 200
-    max: 1000
+  regularization_params.dropout:
```
Reintroduced in a40eca8.

danielzuegner commented on examples/README.md:
```diff
@@ -151,14 +134,18 @@ The special 'slurm' block contains the slurm parameters. This block and all valu
-If a specific configuration (e.g. `large_datasets`) defines the same parameters, they will override the ones defined in the root, e.g. the learning rate in the example above.
-This means that for all configurations in the `large_datasets` the learning rate will be `0.001` and not `0.01` or `0.05` as defined in the root of the document.
+If a specific configuration (e.g. `large_datasets`) defines the same parameters **in the same block** (`fixed`, `random`, 
+`grid`), they will override the ones defined before, e.g. the learning rate in the example above.
```
Fixed in a40eca8.

danielzuegner commented on examples/advanced_example_experiment.py:
```diff
@@ -0,0 +1,96 @@
+        return results
+
+
+@ex.command(unobserved=True)
```
Added some comments in a40eca8 :)

klicperajo reviewed (approved):
Looks great! :)
